R/mock2.R

```diff
@@ -0,0 +1,96 @@
+#' Mocking tools
+#'
+#' @description
+#' `r lifecycle::badge("experimental")`
+#'
+#' These functions represent a second attempt at bringing mocking to testthat,
+#' incorporating what we've learned from mockr and mockery There are
+#' two styles of mocking:
+#'
+#' * `local_mocked_bindings()` and `local_mocked_bindings()` work by temporarily
+#'   changing variable bindings in the package namespace. It affects the
+#'   operation of every function in the current package, but only the current
+#'   package.
+#'
+#' * `with_mocked_env()` modifies a function, inserting a new environment in the
+#'   chain of parents with the specified bindings. It can mock any function,
+#'   but because it creates a new function, it doesn't affect any functions
+#'   that use that function.
+#'
+#' @export
+#' @param f A function.
+#' @param ... Name-value pairs providing functions to mock.
+with_mocked_env <- function(f, ...) {
+  if (!is.function(f)) {
+    cli::cli_abort("{.arg f} must be a function, not {.obj_type_friendly {f}}.")
+  }
+  bindings <- list2(...)
+  check_bindings(bindings)
+
+  new_env <- child_env(get_env(f), !!!bindings)
+  set_env(f, new_env)
+}
+
+#' @export
+#' @rdname with_mocked_env
+#' @param code Code to execute with specified bindings.
+with_mocked_bindings <- function(code, ...) {
+  quo <- enquo(code)
+  local_mocked_bindings(...)
+  code
+}
+
+#' @export
+#' @param .env Environment that defines effect scope. For expert use only.
+#' @rdname with_mocked_env
+local_mocked_bindings <- function(..., .env = caller_env()) {
+  bindings <- list2(...)
+  check_bindings(bindings)
+
+  ns_env <- ns_env(dev_package())
+
+  # Unlock bindings, if needed
+  nms <- names(bindings)
+  locked <- env_binding_are_locked(ns_env, nms)
+  env_binding_unlock(ns_env, nms[locked])
```

Comment:
`env_binding_unlock()` returns the old state, so the `env_binding_are_locked()` call above can be simplified.